File: go/2022/03/solution.go
package main

import (
	"aoc/utils"
	"aoc/utils/set"
)

func main() {
	lines := utils.ReadLines()

	// part 1
	acc := 0
	for _, line := range lines {
		if line == "" {
			continue
		}
		middle := len(line) / 2
		left, right := line[:middle], line[middle:]
		common := set.New([]rune(left)).Intersection(set.New([]rune(right)))
		if len(*common) != 1 {
			panic(len(*common))
		}
		acc += score(common.Pop())
	}

	println(acc)

	// part 2
	acc = 0
	for i := 0; i < len(lines); i += 3 {
		s1 := set.New([]rune(lines[i]))
		s2 := set.New([]rune(lines[i+1]))
		s3 := set.New([]rune(lines[i+2]))
		common := s1.Intersection(s2).Intersection(s3)
		if len(*common) != 1 {
			panic(len(*common))
		}
		acc += score(common.Pop())
	}

	println(acc)
}

func score(letter rune) int {
	if letter >= 'a' {
		return int(letter - 'a' + 1)
	} else {
		return int(letter - 'A' + 27)
	}
}

File: python/2022/03/solution.py
import common

lines = common.read_file().splitlines()

def score_letter(letter):
    if ord(letter) >= ord('a'):
        return ord(letter) - ord('a') + 1 
    else:
        return ord(letter) - ord('A') + 27

# part 1
acc = 0
for line in lines:
    middle = len(line)//2
    left, right = line[:middle], line[middle:]
    common = set(left).intersection(set(right))
    assert len(common) == 1
    acc += score_letter(common.pop())

print(acc)

# part 2
acc = 0
for i in range(0, len(lines), 3):
    common = set(lines[i]).intersection(set(lines[i+1])).intersection(set(lines[i+2]))
    assert len(common) == 1
    acc += score_letter(common.pop())

print(acc)
